src/mono/mono/component/marshal-ilgen-stub.c

```diff
@@ -14,6 +14,37 @@ stub_emit_marshal_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
 		MonoMarshalSpec *spec, int conv_arg,	
 		MonoType **conv_arg_type, MarshalAction action, MonoMarshalLightweightCallbacks* lightweight_cb)
 {
+	if (spec && spec->native == MONO_NATIVE_CUSTOM)
+		return conv_arg;
+
+	if (spec && spec->native == MONO_NATIVE_ASANY)
+		return conv_arg;
+
+	switch (t->type) {
+	case MONO_TYPE_BOOLEAN:
+		return lightweight_cb->emit_marshal_scalar (m, argnum, t, spec, conv_arg, conv_arg_type, action);
+	case MONO_TYPE_PTR:
+		return lightweight_cb->emit_marshal_scalar (m, argnum, t, spec, conv_arg, conv_arg_type, action);
+	case MONO_TYPE_CHAR:
+		return lightweight_cb->emit_marshal_scalar (m, argnum, t, spec, conv_arg, conv_arg_type, action);
+	case MONO_TYPE_I1:
+	case MONO_TYPE_U1:
+	case MONO_TYPE_I2:
+	case MONO_TYPE_U2:
+	case MONO_TYPE_I4:
+	case MONO_TYPE_U4:
+	case MONO_TYPE_I:
+	case MONO_TYPE_U:
+	case MONO_TYPE_R4:
+	case MONO_TYPE_R8:
+	case MONO_TYPE_I8:
+	case MONO_TYPE_U8:
+	case MONO_TYPE_FNPTR:
+		return lightweight_cb->emit_marshal_scalar (m, argnum, t, spec, conv_arg, conv_arg_type, action);
+	default:
+		return conv_arg;
```

Comment:
I wonder if we can assert here.  (I guess ideal would be to emit a method body that throws an exception when it is called, but that might be challenging to do).

We don't get here for icalls - those end up in `emit_native_icall_wrapper_ilgen`, so in theory we shouldn't see strings or arrays or objects here.

We also don't get here if the runtime marshaling disabled attribute is applied.  in that case we go to `mono_emit_disabled_marshal` which can handle blittable structs.

so if we see a valuetype here (or a generic instance that is a valuetype), even if it's blittable we probably should assert.

Reply:
I'll see if I can throw an exception here, it makes sense to me.